python/ray/dag/dag_node.py

```diff
@@ -58,12 +58,30 @@ def __init__(
         self._bound_other_args_to_resolve: Optional[Dict[str, Any]] = (
             other_args_to_resolve or {}
         )
+
+        # The list of nodes that use this DAG node as input.
+        self._downstream_nodes: List[DAGNode] = []
```

Comment:
will there be any circular dependency in this case? maybe we should make it contain a weakref

Reply:
It works on CI and my devbox. 

> maybe we should make it contain a weakref

Do you mean use `List["DAGNode"]` instead?

Reply:
Updated https://github.com/ray-project/ray/pull/47261/commits/54df435b62b4510d731bd9df4b766f694e54d5ea

Reply:
yeah iiuc downstream node has a reference to upstream node and vice versa. So we should make one of the contain the weakref otherwise it will have circulare ref 

For exmaple, when you do
```
            upstream_node._downstream_nodes.append(self)

```

it could be instead

```
            upstream_node._downstream_nodes.append(weakref(self))

```

(weakref is not reference counted)

Reply:
I wrote a summary about circular reference: https://github.com/ray-project/ray/pull/47261#issuecomment-2306118753